src/bin/miri.rs

```diff
@@ -63,6 +64,34 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
     }
 }
 
+struct OverrideExportedSymbols {
+    target_crate: bool,
+}
+
+impl rustc_driver::Callbacks for OverrideExportedSymbols {
+    fn config(&mut self, config: &mut Config) {
+        if config.opts.prints.is_empty() && self.target_crate {
+            config.override_queries = Some(|_, local_providers, _| {
+                local_providers.exported_symbols = |tcx, cnum| {
+                    assert_eq!(cnum, LOCAL_CRATE);
+                    tcx.arena.alloc_from_iter(
+                        tcx.reachable_set(LOCAL_CRATE).iter().filter_map(|&local_def_id| {
+                            tcx.codegen_fn_attrs(local_def_id)
+                                .contains_extern_indicator()
+                                .then_some((
+                                    ExportedSymbol::NonGeneric(local_def_id.to_def_id()),
+                                    SymbolExportLevel::C,
+                                ))
+                        })
+                    )
+                }
+            });
+            // FIXME: This seems required to let libstd get rid of `panic_unwind`.
```

Comment:
What do you mean by "get rid of"?

Reply:
libstd seems to bundle a panic runtime if it's "built" as a `dylib`.

Reply:
Okay... I still don't understand what you mean by "get rid off" though. Which problem is this solving?

Reply:
Well this is solving the `__rust_start_panic` problem above: https://github.com/rust-lang/miri/pull/1776#discussion_r615421604

Reply:
Ah. :-) So please say that -- say something about how we end up with two definitions of `__rust_start_panic` in the crate tree if we don't do this.

But doesn't this indicate something is more fundamentally wrong? There shouldn't be two exports of the same symbol, shouldn't rustc detect that? Won't that cause other problems?

Reply:
I guess that `rustc` will first filter out both `panic_unwind` and `panic_abort` in https://github.com/rust-lang/rust/blob/e7ed7e835494219eeab314ace8beef40b695ce30/compiler/rustc_metadata/src/dependency_format.rs#L290. and then add only `panic_abort` back in https://github.com/rust-lang/rust/blob/e7ed7e835494219eeab314ace8beef40b695ce30/compiler/rustc_metadata/src/dependency_format.rs#L298-L300...
I suspect that code is related to [`TyCtxt::dependency_formats`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.dependency_formats), so I think I could try to use that instead of `TyCtxt::crates`.

Reply:
> I suspect that code is related to [`TyCtxt::dependency_formats`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.dependency_formats), so I think I could try to use that instead of `TyCtxt::crates`.

No, that does not work. :cry: I got a bunch of ["nice pretty errors"](https://github.com/rust-lang/rust/blob/e7ed7e835494219eeab314ace8beef40b695ce30/compiler/rustc_metadata/src/dependency_format.rs#L126):

<details>

```
+error: crate `std` required to be available in rlib format, but was not found in this form
+
+error: crate `core` required to be available in rlib format, but was not found in this form
+
+error: crate `compiler_builtins` required to be available in rlib format, but was not found in this form
+
+error: crate `rustc_std_workspace_core` required to be available in rlib format, but was not found in this form
+
+error: crate `alloc` required to be available in rlib format, but was not found in this form
+
+error: crate `libc` required to be available in rlib format, but was not found in this form
+
+error: crate `unwind` required to be available in rlib format, but was not found in this form
+
+error: crate `cfg_if` required to be available in rlib format, but was not found in this form
+
+error: crate `hashbrown` required to be available in rlib format, but was not found in this form
+
+error: crate `rustc_std_workspace_alloc` required to be available in rlib format, but was not found in this form
+
+error: crate `rustc_demangle` required to be available in rlib format, but was not found in this form
+
+error: crate `std_detect` required to be available in rlib format, but was not found in this form
+
+error: crate `panic_unwind` required to be available in rlib format, but was not found in this form
```
</details>

Reply:
This code has gone now since I replaced `TyCtxt::crates()` with `TyCtxt::dependency_formats()`. `dependency_formats()` handles `panic_unwind`+`panic_abort` gracefully, although I have to workaround the "crate ... required to be available in rlib format, but was not found in this form" errors.